TheAlienThatEatsTheCarrot/GameEngine/Systems/GameSystems/GameEndSystem.swift

```diff
@@ -0,0 +1,63 @@
+//
+//  GameEndSystem.swift
+//  TheAlienThatEatsTheCarrot
+//
+//  Created by Sun Xinyu on 15/4/24.
+//
+
+import Foundation
+
+class GameEndSystem: System {
+    var nexus: Nexus
+    weak var gameEngine: GameEngine?
+    weak var gameEndObserver: NSObjectProtocol?
+
+    init(nexus: Nexus, gameEngine: GameEngine?) {
+        self.nexus = nexus
+        self.gameEngine = gameEngine
+        subscribeToEvents()
+    }
+
+    func subscribeToEvents() {
+        gameEndObserver = EventManager.shared.subscribe(to: GameEndEvent.self, using: onGameEnd)
+    }
+
+    private lazy var onGameEnd = { [weak self] (event: Event) -> Void in
+        guard let strongSelf = self,
+              let gameEndEvent = event as? GameEndEvent else {
+            return
+        }
+        strongSelf.gameEngine?.end()
+    }
+
+    func update(deltaTime: CGFloat) {
+        checkWinCondition()
+        checkLoseCondition()
+    }
+
+    private func checkWinCondition() {
```

Comment:
Do we want to make this extensible? Suppose there are different "game modes" in the future with different win/lose conditions. We want to be able to accommodate that. I think I good solution is to use the strategy pattern. Where the GameEndSystem will have a reference to a `GameMode` or something. And the `NormalGameMode` implements its own `checkWinCondition` and `checkLoseCondition`. So the new code would be something like:
```swift
func update(deltaTime: CGFloat, gameMode: GameMode) {
  gameMode.checkWinCondition()
  gameMode.checkLoseCondition()
}
```
The implementation details can be decided. Eg. Game mode can be a component or something etc. but the idea stays the same, we can create new game modes in the future with different win lose conditions. Eg. instead of win when collecting 3 carrots, it wins when all enemies are killed. Instead of lose when lives are out, it loses when time runs out etc.